src/us/kbase/workspace/test/database/provenance/ResolvablePIDTest.java

```diff
@@ -0,0 +1,137 @@
+		assertThat(INCORRECT_REL_TYPE, rpid1.getRelationshipType(), is(Optional.empty()));
+	}
+
+	@Test
+	public void buildMaximal() throws Exception {
+		final ResolvablePID rpid1 = ResolvablePID.getBuilder(RPID_STRING).withDescription(DESC_STRING)
+				.withRelationshipType(REL_STRING).build();
+		assertThat(INCORRECT_RPID, rpid1.getId(), is(RPID_STRING));
+		assertThat(INCORRECT_DESC, rpid1.getDescription(), is(opt(DESC_STRING)));
+		assertThat(INCORRECT_REL_TYPE, rpid1.getRelationshipType(), is(opt(REL_STRING)));
+	}
+
+	@Test
+	public void buildTrimAllFields() throws Exception {
+		final ResolvablePID rpid1 = ResolvablePID.getBuilder(RPID_STRING_UNTRIMMED)
+				.withDescription(DESC_STRING_UNTRIMMED).withRelationshipType(REL_STRING_UNTRIMMED).build();
+		assertThat(INCORRECT_RPID, rpid1.getId(), is(RPID_STRING));
+		assertThat(INCORRECT_DESC, rpid1.getDescription(), is(opt(DESC_STRING)));
+		assertThat(INCORRECT_REL_TYPE, rpid1.getRelationshipType(), is(opt(REL_STRING)));
+	}
+
+	@Test
+	public void buildWithWhitespaceDescriptionRelationshipType() throws Exception {
+		final ResolvablePID rpid1 = ResolvablePID.getBuilder(RPID_STRING_UNTRIMMED).withDescription(WHITESPACE)
+				.withRelationshipType(WHITESPACE).build();
+		assertThat(INCORRECT_RPID, rpid1.getId(), is(RPID_STRING));
+		assertThat(INCORRECT_DESC, rpid1.getDescription(), is(Optional.empty()));
+		assertThat(INCORRECT_REL_TYPE, rpid1.getRelationshipType(), is(Optional.empty()));
+	}
+
+	@Test
+	public void buildWithNullDescriptionRelationshipType() throws Exception {
+		final ResolvablePID rpid1 = ResolvablePID.getBuilder(RPID_STRING).withDescription(null)
+				.withRelationshipType(null).build();
+		assertThat(INCORRECT_RPID, rpid1.getId(), is(RPID_STRING));
+		assertThat(INCORRECT_DESC, rpid1.getDescription(), is(Optional.empty()));
+		assertThat(INCORRECT_REL_TYPE, rpid1.getRelationshipType(), is(Optional.empty()));
+	}
+
+	@Test
+	public void buildAndOverwriteDescriptionRelationshipType() throws Exception {
+		final ResolvablePID rpid1 = ResolvablePID.getBuilder(RPID_STRING).withDescription(DESC_STRING)
+				.withDescription(WHITESPACE).withRelationshipType(REL_STRING).withRelationshipType(WHITESPACE).build();
+		assertThat(INCORRECT_RPID, rpid1.getId(), is(RPID_STRING));
+		assertThat(INCORRECT_DESC, rpid1.getDescription(), is(Optional.empty()));
+		assertThat(INCORRECT_REL_TYPE, rpid1.getRelationshipType(), is(Optional.empty()));
+
+		final ResolvablePID rpid2 = ResolvablePID.getBuilder(RPID_STRING).withDescription(DESC_STRING)
+				.withDescription(null).withRelationshipType(REL_STRING).withRelationshipType(null).build();
+		assertThat(INCORRECT_RPID, rpid2.getId(), is(RPID_STRING));
+		assertThat(INCORRECT_DESC, rpid2.getDescription(), is(Optional.empty()));
+		assertThat(INCORRECT_REL_TYPE, rpid2.getRelationshipType(), is(Optional.empty()));
+	}
+
+	@Test
+	public void buildFailInvalidRPID() throws Exception {
+		try {
```

Comment:
I leave this up to you, but what I would do here is make positive and negative test functions and run a list of positive and negative RPIDs through them with a range of characters and formats

Reply:
:+1: